Fix MOTD not showing up on tsh login with certain arguments

lxea opened the pull request:
- changes to configuration.go: fixes tsh login in the first case from the repro files
  `tsh login --insecure --proxy=127.0.0.1:3080 --user=test`
- changes to apiserver.go fixes `--auth` and `--with-roles` not showing motd

These changes also add a test and part of that required moving `export_test.go` to `export.go` so the `passwordFromConsole` function could be overridden when testing for the MOTD

fixes: #10503

Commit:
Fix MOTD not showing up on tsh login with certain arguments

- changes to configuration.go: fixes tsh login in first test case
  `tsh login --insecure --proxy=127.0.0.1:3080 --user=test`
- changes to apiserver.go fixes `--auth` not showing motd

r0mant commented on lib/client/api.go:
```diff
@@ -2390,14 +2390,8 @@ func (tc *TeleportClient) LogoutAll() error {
-func (tc *TeleportClient) Login(ctx context.Context) (*Key, error) {
-	// Ping the endpoint to see if it's up and find the type of authentication
-	// supported.
+// PingAndShowMOTD makes a ping aisplays the MOTD.
```
```suggestion
// PingAndShowMOTD pings the Teleport Proxy and displays the MOTD if it's available.
```

r0mant commented on lib/config/configuration.go:
```diff
@@ -547,8 +547,8 @@ func applyAuthConfig(fc *FileConfig, cfg *service.Config) error {
 		}
-		cfg.Auth.Preference.SetMessageOfTheDay(fc.Auth.MessageOfTheDay)
 	}
+	cfg.Auth.Preference.SetMessageOfTheDay(fc.Auth.MessageOfTheDay)
```
Should this only be set in case `fc.Auth.MessageOfTheDay != ""`?

r0mant commented on lib/web/apiserver.go:
```diff
@@ -788,33 +788,40 @@ func (h *Handler) pingWithConnector(w http.ResponseWriter, r *http.Request, p ht
+		authPresent = true
+	}
+
+	if authPresent {
+		as.HasMessageOfTheDay = cap.GetMessageOfTheDay() != ""
+		response.Auth = as
 		return response, nil
 	}
```
If you kept the old logic of assigning `response.Auth = xxx` directly above, would you be able to get rid of the extra `authPresent` boolean flag and just use `if response.Auth != nil` check instead? Then you could something like
```suggestion
	if response.Auth != nil {
		response.Auth.HasMessageOfTheDay = cap.GetMessageOfTheDay() != ""
	}
```

which is a bit cleaner.

lxea replied:
response.Auth isn't nilable at the moment, should I make it a pointer in this pr?

r0mant commented on tool/tsh/tsh.go:
```diff
@@ -864,6 +868,10 @@ func onLogin(cf *CLIConf) error {
 				return trace.Wrap(err)
 			}
 			printProfiles(cf.Debug, profile, profiles)
+			_, err := tc.PingAndShowMOTD(cf.Context)
```
Hmm, you already have PingAndShowMOTD above, aren't we going to display it twice if you also put it in each "case" branch here?

lxea replied:
Ah yeah, I meant to remove that, it could also print twice being at the top if it flows through too.

lxea replied:
I did it like this to retain the previous `tc.Login` behavior -- it shows the MOTD when its called and uses the result of the ping. I can change it so showing the MOTD is done separatly and the ping is passed in

r0mant replied:
@lxea Ah yes, it makes sense. In that case it should be fine.

r0mant replied:
But can you please move it so MOTD is displayed before everything else like profiles?

r0mant reviewed (commented):
Please also include test coverage. We have some tsh tests already, should be pretty straightforward.

Commit:
Add tests for motd fixes

Part of this includes renaming export_test.go to export.go so I could
test the MOTD outside of lib/client/export.go

r0mant commented:
@jakule or @zmb3 can one of you folks take a look as well pls?

jakule commented on lib/client/api.go:
```diff
@@ -2390,14 +2390,8 @@ func (tc *TeleportClient) LogoutAll() error {
-func (tc *TeleportClient) Login(ctx context.Context) (*Key, error) {
-	// Ping the endpoint to see if it's up and find the type of authentication
-	// supported.
+// PingAndShowMOTD pings the Teleport Proxy and displays the MOTD if it's available.
```
nit: I don't abbreviation in the function name, but I'd assume that comment explains what MOTD is. Without any context this can be not obvious.
```suggestion
// PingAndShowMOTD pings the Teleport Proxy and displays the Message Of The Day if it's available.
```

jakule commented on tool/tsh/tsh_test.go:
```diff
@@ -231,11 +240,24 @@ func TestOIDCLogin(t *testing.T) {
+			break
+		}
+	}
+	require.True(t, found, "Failed to find %q MOTD in the logs", motd)
```
nit: I don't think `found`variable is needed here. You could replace the break a few lines above with `return` and this line with `require.Fail`.  In my opinion that would simplify this function.

jakule commented on tool/tsh/tsh_test.go:
```diff
@@ -1252,7 +1294,12 @@ func withBootstrap(bootstrap ...types.Resource) testServerOptFunc {
 func withAuthConfig(fn func(cfg *service.AuthConfig)) testServerOptFunc {
 	return func(o *testServersOpts) {
-		o.authConfigFunc = fn
+		fmt.Println("honk1", o.authConfigFuncs)
```
This looks like some debug leftover.

jakule reviewed (approved):
One comment, two nits.
